services/brig/src/Brig/SMTP.hs

```diff
@@ -72,18 +140,85 @@ initSMTP lg host port credentials connType = do
           SMTP.connectSMTPSSLWithSettings (unpack host) $
             SMTP.defaultSettingsSMTPSSL {SMTP.sslPort = p}
       ok <- case credentials of
-        (Just (Username u, Password p)) -> SMTP.authenticate SMTP.LOGIN (unpack u) (unpack p) conn
+        (Just (Username u, Password p)) ->
+          liftSMTP $
+            SMTP.authenticate SMTP.LOGIN (unpack u) (unpack p) conn
         _ -> pure True
-      pure (ok, conn)
-    create = do
-      (ok, conn) <- connect
       if ok
-        then Logger.log lg Logger.Debug (msg $ val "Established connection to: " +++ host)
-        else Logger.log lg Logger.Warn (msg $ val "Failed to established connection, check your credentials to connect to: " +++ host)
-      pure conn
-    destroy c = do
-      SMTP.closeSMTP c
-      Logger.log lg Logger.Debug (msg $ val "Closing connection to: " +++ host)
-
-sendMail :: MonadIO m => SMTP -> Mail -> m ()
-sendMail s m = liftIO $ withResource (s ^. pool) $ SMTP.sendMail m
+        then pure conn
+        else throwE Unauthorized
+
+    create :: IO SMTP.SMTPConnection
+    create = do
+      res <- runExceptT establishConnection
+      logResult lg ("Creating pooled SMTP connection to " ++ unpack host) res
+      throwOnLeft res
+
+    destroy :: SMTP.SMTPConnection -> IO ()
+    destroy c =
+      (ensureSMTPConnectionTimeout timeoutDuration . SMTP.gracefullyCloseSMTP) c
+        >>= void . logResult lg ("Closing pooled SMTP connection to " ++ unpack host)
+
+throwOnLeft :: MonadIO m => Either SMTPFailure a -> m a
+throwOnLeft = \case
+  Left Unauthorized -> CE.throw SMTPUnauthorized
+  Left ConnectionTimeout -> CE.throw SMTPConnectionTimeout
+  Left (CaughtException e) -> CE.throw e
+  Right a -> pure a
+
+logResult :: MonadIO m => Logger -> String -> Either SMTPFailure c -> m ()
+logResult lg actionString res =
+  case res of
+    Left Unauthorized -> do
+      Logger.log
+        lg
+        Logger.Warn
+        (msg $ concatToVal actionString "Failed to establish connection, check your credentials.")
+    Left ConnectionTimeout -> do
+      Logger.log lg Logger.Warn (msg $ concatToVal actionString "Connection timeout.")
+    Left (CaughtException e) -> do
+      Logger.log lg Logger.Warn (msg $ concatToVal actionString ("Caught exception : " ++ show e))
+    Right _ -> do
+      Logger.log lg Logger.Debug (msg $ concatToVal actionString "Succeeded.")
+  where
+    concatToVal :: ToBytes s1 => s1 -> String -> Builder
+    concatToVal a b = a +++ (" : " :: String) +++ b
```

Comment:
this looks a little like a reinvention of log entry fields.  but i'm find with that as long as it's local, and we aren't likely to want to use the structure of the entry in dashboards for understanding the data better.

Reply:
I've changed the implementation to use `fields`.